### ISSUE TITLE
Support programmatically changing focus

### ISSUE THREAD

**Issue by mfs780** [Workday/canvas-kit#618]
## 🐛 Bug Report

Currently when focus is changed to an input programmatically via .focus() the blue ring is hidden. This is due to a CSS override caused by data-whatinput="mouse" :focus

## To Reproduce

Utilize .focus on a selected TextButton.

## Expected Behavior

When programmatically focusing on a TextButton the blue focus ring should show.

## Actual Results

There is no blue ring.

### Error Output

Include a screenshot of a visual bug

<img width="314" alt="Screen Shot 2020-05-04 at 3 39 08 PM" src="https://user-images.githubusercontent.com/4986491/81092113-fb530180-8eb4-11ea-9e55-05c2e65f467f.png">

**Comment by NicholasBoll** [comment]
Some ideas on what to do here:

### Allow programatic change of input type
This is the easiest for us because there is little to support. The burden is on everyone else to call an update. This could be accomplished by exposing a React Context that is provided by the InputProvider to allow overriding of the input type ("programmatic" could be a type meaning focus changed outside a direct user-action to do so). It could also be accomplished through a simple function call - as of v4 all InputProvider side-effects are outside the React lifecycle anyway so React Context isn't actually needed.

Pros:
- Almost no maintenance cost
- Application developers are responsible for any edge cases

Cons:
- Everyone has to opt-in to this override for each instance
- Increases burden on application developers to handle the decision to hide focus rings on mouse clicks made by the Design System teams

### We automatically detect programmatic focus changes
This is more difficult for us because we have to support edge cases where such detection fails. Some of these failures would be indirect and hard to trace. The benefit is it requires no extra work from others to "just do the right thing". Programmatic detection is possible through some state tracking. For instance, an input getting focus from a mouse is the default action of a `mousedown` event (`event.preventDefault()` will prevent focus changes). We can detect the element that was fired from a `focus` and see if it is the same one from a `mousedown` or `touchstart, `pointerdown`, etc. If the elements don't match, it was from a programmatic focus change.

Pros:
- No work required for application developers. It is easy to follow the design system guidelines
- Automatic consistency

Cons:
- More work on us. There may be many edge cases that need to be ironed out
- InputProvider does more things on more global events - we have to be careful not to slow down the application with extra state tracking

**Comment by NicholasBoll** [comment]
For automatic detection, is there opposition to overriding `HTMLElement.prototype.focus` to include a call to change detected input device?

```ts
const nativeFocus = HTMLElement.prototype.focus

HTMLElement.prototype.focus = function (...args) {
  console.log('focus changed'); // change detection here

  nativeFocus.apply(this, args)
}
```

This would only be called when someone explicitly calls `.focus()` on an element and doesn't override the `focus` method (that would only be problem on custom elements with custom focus methods).

**Comment by anicholls** [comment]
`InputProvider` works by setting some data attributes on its `container` element (the `body` element by default):

```html
<body data-whatinput="mouse" data-whatinput="mouse">
```

As a workaround, you should be able to trick it to show focus by manually setting these to `"keyboard"` before calling `.focus()`

**Comment by NicholasBoll** [comment]
Closed via #1065

```tsx
import {changeFocus} from '@workday/canvas-kit-react/common';

changeFocus(element)
```
